src/Maestro/DependencyUpdateErrorProcessor/DependencyUpdateErrorProcessor.cs

```diff
@@ -0,0 +1,386 @@
+                            tx,
+                            "update",
+                            error.Timestamp
+                        );
+                        await tx.CommitAsync();
+                    }
+                }
+                catch (Exception ex)
+                {
+                    Logger.LogError(ex,$"Unable to create a github issue for error message : '{error.ErrorMessage}' for the repository : '{error.Repository}'");
+                }
+            }
+        }
+
+        private async Task<GitHubClient> AuthenticateGitHubClient(string issueRepo)
+        {
+            IGitHubTokenProvider gitHubTokenProvider = Context.GetService<IGitHubTokenProvider>();
+            var installationId = await Context.GetInstallationId(issueRepo);
+            var gitHubToken = await gitHubTokenProvider.GetTokenForInstallationAsync(installationId);
+            Logger.LogInformation($"GitHub token acquired for '{issueRepo}'");
+            var version = Assembly.GetExecutingAssembly()
+                .GetCustomAttribute<AssemblyInformationalVersionAttribute>()
+                .InformationalVersion;
+            Octokit.ProductHeaderValue product = new Octokit.ProductHeaderValue("Maestro", version);
+            return new GitHubClient(product)
+            {
+                Credentials = new Credentials(gitHubToken),
+            };
+        }
+
+        private async Task CreateOrUpdateGithubIssueAsync(RepositoryBranchUpdateHistoryEntry updateHistoryError, string issueRepo)
+        {
+            Logger.LogInformation($"Error Message : '{updateHistoryError.ErrorMessage}' in repository :  '{updateHistoryError.Repository}'");
+            var gitHubIssueEvaluator =
+                await _stateManager.GetOrAddAsync<IReliableDictionary<(string, string), int>>("gitHubIssueEvaluator");
+            var client = await AuthenticateGitHubClient(issueRepo);
+            var repo = await client.Repository.Get(_options.Owner, _options.Repository);
+            var issueNumber = new ConditionalValue<int>();
+            using (ITransaction tx = _stateManager.CreateTransaction())
+            {
+                issueNumber = await gitHubIssueEvaluator.TryGetValueAsync(tx,
+                    (updateHistoryError.Repository, updateHistoryError.Branch));
+                await tx.CommitAsync();
+            }
+            if (issueNumber.HasValue)
+            {
+                Logger.LogInformation(
+                    $"Updating a gitHub issue number : '{issueNumber}' for the error : '{updateHistoryError.ErrorMessage}'"
+                    + $"for the repository : '{updateHistoryError.Repository}'");
+                var issue = await client.Issue.Get(repo.Id, issueNumber.Value);
+                var descriptionUpdateFlag = false;
+                var comments = await client.Issue.Comment.GetAllForIssue(repo.Id, issue.Number);
+                var updateIssueDescription = CreateOrUpdateDescription(updateHistoryError, issue.Body);
```

Comment:
I think I'm not understanding why this has to be called multiple times. You're always building the same description regardless, and you just need to find whether it should be added to the body or to the comments, or create a new comment?

Reply:
I think you can just build the description regardless, and then the method only needs to return whether it should go in the body, in an existing comment, or in a new comment.

Reply:
And depending on where it goes, you add the additional stuff like the FYI handles

Reply:
I am calling it once before Create / Update and once for every comment ( if there are any comments) 

The reason I call it multiple times in the comments, cos I need to check each comment if the error already exists, if it does then I have to replace the body. 

Let me know I can explain further .

Reply:
Yep I understand that part, it just seems odd, since you're repeating a bunch of operations that don't need to be repeated, because you need to search for the message.

Reply:
Building description is different for different methods. Like for updateAssestsAsync I will have subscriptionid, so I can get source and target repository but for ProcessPending, I will have generic details. Also I combined 2 different methods to one so that I don't have to explictly check for methods twice.

Reply:
You are calling the method in three different places and the only thing that is different is the place where you're searching for the message. This means that some of the logic is going to execute three times unnecessarily (including the switch statement).

* https://github.com/dotnet/arcade-services/pull/1041/files#diff-cba256505839b1a8a00717979aa29a2cR163
* https://github.com/dotnet/arcade-services/pull/1041/files#diff-cba256505839b1a8a00717979aa29a2cR180
* https://github.com/dotnet/arcade-services/pull/1041/files#diff-cba256505839b1a8a00717979aa29a2cR207

This tells me that the method could take some extra parameters, like the built description, and the function that needs to be used to find the message.

Reply:
Do you suggest a better way? I tried to simplify it as much as I could. :(

Reply:
I did suggest what I believe is a better way here: https://github.com/dotnet/arcade-services/pull/1041#discussion_r391982746

to summarize, 

* You can move the switch to [here](https://github.com/dotnet/arcade-services/pull/1041/files#diff-cba256505839b1a8a00717979aa29a2cR106)

* Inside the switch, you build the description, and you declare a `Func<string, bool>` object. This is a function that takes a string (the text inside an issue body or a comment) and returns a boolean (whether it was found or not)
* You build the description for the error here as well.
* CreateOrUpdateGithubIssueAsync now takes two additional paramenters, the description you want to add, and the function object.
* CreateOrUpdateGithubIssueAsync becomes much simpler, as you iterate over the body and comments, running the function that was passed as a parameter, and inserting the error message where it's supposed to.